text/0000-compose.md

```diff
@@ -0,0 +1,125 @@
+
+```js
+import {composePlugins} from 'fusion-core';
+function composeSum(items) {
+  return items.reduce((sum, item) => {
+    return sum + item;
+  }, 0);
+}
+app.register(ComposeExampleToken, composePlugins([
+  createPlugin({
+    provides: () => 2
+  }),
+  createPlugin({
+    provides: () => 3
+  }),
+]), composeSum);
+
+// ComposeExampleToken => 5
+```
+
+# Motivation
+
+There are often situations where you need to manage composing a large amount of objects onto a single token. This can be done today with a few approaches today,
+but they all require significant boilerplate and often unreasonable overheard. For example, you may want to split your graphql schemas and resolvers up into many
+separate modular schemas, and combine them onto a single token using schema stiching. However as the number of separate schemas expands upwards of 10, 20, it 
+becomes restrictive to manage separate tokens for each separate schema.
+
+## Use Cases
+
+A real life use case of `composePlugins` would be managing graphql schemas.
+
+```js
+const TokenA = createToken('TokenA');
+const TokenB = createToken('TokenB');
+const TokenC = createToken('TokenC');
+const TokenD = createToken('TokenD');
+const TokenE = createToken('TokenE');
+// Old way
+app.register(TokenA, SchemaA);
+app.register(TokenB, SchemaB);
+app.register(TokenC, SchemaC);
+app.register(TokenD, SchemaD);
+app.register(TokenE, SchemaE);
+
+app.register(SchemaToken, createPlugin({
+  deps: {
+    a: TokenA,
+    b: TokenB,
+    c: TokenC,
+    d: TokenD,
+    e: TokenE,
+  },
+  provides({a, b, c, d, e}) {
+    return mergeSchemas([a,b,c,d,e])
+  }
+}))
+```
+
+This becomes exceedingly difficult as the number of schemas increases. With the app.compose method, this could be replaced with:
+
+```js
+// new way
+app.register(SchemaToken, composePlugins([SchemaA, SchemaB, SchemaC, SchemaD, SchemaE], composeFn));
+```
+
+Another real life use case could be managing redux store enhancers.
+
+```js
+app.register(ReduxEnhancerToken, composePlugins([enhancerA, enhancerB, enhancerC], composeFn)
+```
+
+# Detailed design
+
+This RFC proposes a new export from `fusion-core`: `composePlugins`. The `composePlugins` function will take a list of plugins and a compose function which
+will be used to compose their resolved values. 
+
+The type definition will look something like this:
+```js
+type composePlugins<T> = (plugins: Array<FusionPlugin<any, T>, composeFn: (values: Array<T>) => T): FusionPlugin<any, T>;
+```
+
+# Drawbacks
+
+This approach still requires a centralized place where plugins are composed. This can be somewhat restrictive when working with a large number of plugins that need
+to be composed (upwards of 50 to 100). Having a API which allows distributed composition provides more flexibility when managing code organization. For example, it
+would support organizing by feature rather than by type. However, this rigidity could also provide benefits to readability and debuggability. 
+
```

Comment:
Per https://github.com/fusionjs/rfcs/pull/14#issuecomment-474555494, a decentralized approach is still theoretically possible. Should we show an example of how this could be done or what this would look like?

I suppose we should compare the drawbacks of this extra layer on top vs. integrating into the core API.

Reply:
For posterity, I'm pasting a snippet below from a private convo, which illustrates feature parity between app.compose and createComposedPlugin in terms of decentralized compose calls

```
//sugar
const comps = new Map()
app.compose = (key, value) => (comps.get(key) || comps.set(key, []).get(key)).push(value)

//usage
app.compose(A, a)
app.register(A, createComposedPlugin({plugins: comps.get(A), ...}))
```

Reply:
For the use cases we have regarding this API; a centralized approach will not work or add any benefit. 
If everything has to be added in one place it is relatively easy to build on top of current fusion. 

```
const schemas = [SchemaPluginA, SchemaPluginB]
app.register(SchemaPlugins, schema)
app.register(SchemaToken, /* plugin that composes the schemaplugins and provides that */)
```

Of course it is a cumbersome way. 

The true power of this api lies in the possibility of being decentralized, which works well in a feature based architecture, in which code that are concerned with the same elements can exist together and does not have to be wired up in a centralized place.

Reply:
I don't think it is possible to have a distributed approach with this API. I don't think the solution of passing around an array is realistic

Reply:
I still think decentralization isn't much of an issue. If we talk purely in terms of what code is required to make this work, at some level of abstraction centralization has to occur in order to bring schemas together, whether that happens w/ `comps` in my snippet above or in a hidden Map deep inside fusion-core. Case in point, if the snippet above _was_ in fusion-core, you'd have "decentralized" `app.compose`. Similarly, a simple monkeypatch of app.register would give you "decentralized" `app.register`

I think the objection is actually less about decentralization and more about explicitness vs implicitness. Having a global Map in app space is explicit, having it tucked away inside fusion-core makes it implicit. Personally, when designing APIs, I prefer to lean on the side of being explicit, lest people will come back later complaining about too much magic or complexity. For example, with my snippet above, it's trivial to test whether you've missed a schema, whereas with state fully encapsulated in fusion-core, it's not.

All that aside, it's also important to consider that an API (especially a _composition_ API) ideally caters to many use cases, and not just a specific one. In other words, composition is a means to an end, not an end in and of itself.

Composed plugins also caters to use cases where a plugin must be registered multiple times, each time with their own dependency trees of configs. We had one such case w/ a dynamic configuration service which we got working w/ a workaround involving aliases and numerous extraneous tokens, but composed plugins would provide a far cleaner API and less Token space pollution.

Overall, I feel like composed plugins are less intrusive implementation-wise, more extensible and more powerful than app.compose.